crates/isograph_compiler/src/write_artifacts.rs

```diff
@@ -1,69 +1,88 @@
-use std::{
-    fs::{self, File},
-    io::Write,
-    path::PathBuf,
+use std::{fs, path::PathBuf};
+
+use common_lang_types::{
+    ArtifactPathAndContent, Diagnostic, DiagnosticResult, FileSystemOperation,
 };
 
-use common_lang_types::{ArtifactPathAndContent, Diagnostic, DiagnosticResult};
-use intern::string_key::Lookup;
+use artifact_content::FileSystemState;
 
 #[tracing::instrument(skip_all)]
-pub(crate) fn write_artifacts_to_disk(
-    paths_and_contents: impl IntoIterator<Item = ArtifactPathAndContent>,
+pub(crate) fn get_file_system_operations(
+    paths_and_contents: &[ArtifactPathAndContent],
     artifact_directory: &PathBuf,
-) -> DiagnosticResult<usize> {
-    if artifact_directory.exists() {
-        fs::remove_dir_all(artifact_directory).map_err(|e| {
-            unable_to_do_something_at_path_diagnostic(
+    file_system_state: &mut Option<FileSystemState>,
+) -> Vec<FileSystemOperation> {
+    let new_file_system_state = FileSystemState::from(paths_and_contents);
+    let operations = match file_system_state {
+        None => {
+            let operations =
```

Comment:
nit: no need to create a variable here or below, just `None => FileSystemState::recreate_all(...)`